okio/jvm/src/main/java/okio/Pipe.kt

```diff
@@ -116,6 +145,40 @@ class Pipe(internal val maxBufferSize: Long) {
     override fun timeout(): Timeout = timeout
   }
 
+  /**
+   * Writes any buffered contents of this pipe to `sink`, then replace this pipe's source with
+   * `sink`. This pipe's source is closed and attempts to read it will throw an
+   * [IllegalStateException].
+   *
+   * This method must not be called while concurrently accessing this pipe's source. It is safe,
+   * however, to call this while concurrently writing this pipe's sink.
+   */
+  @Throws(IOException::class)
+  fun fold(sink: Sink) {
```

Comment:
Just thought of a crazy and ugly edge case . . . what if the `sink.write(...)` call fails ? I think we put things in a very ugly state.

The fix should be really easy. We need a finally clause on the `sink.write()` call that sets `sourceClosed = true` if that method throws

Reply:
```
try {
  sinkBuffer.write(buffer, buffer.size)
} catch (e: IOException) {
  sourceClosed = true
  throw e
}
```
?

Reply:
btw, I `try/finally` inside `mergeOnce` so that we always set the delegate timeout to what it was, even when write/flush/close throws.

Reply:
double checking but you close the source even when it doesn't throw, so it should be inside a catch right?
```kotlin
try {
  sink.write(sinkBuffer, sinkBuffer.size)
} catch (e: IOException) {
  synchronized(buffer) {
    sourceClosed = true
    (buffer as Object).notifyAll() // Notify the sink that it can resume writing.
  }
  throw e
}

```

Reply:
try/finally is a little neater cause it'll handle the case where it's an unchecked exception